horovod/run/run.py

```diff
@@ -714,28 +789,140 @@ def run():
+
+    if args.run_func:
+        run_func_server = KVStoreServer(verbose=settings.verbose)
+        run_func_server_port = run_func_server.start_server()
+        pickled_exec_func = cloudpickle.dumps(args.run_func)
+        put_data_into_kvstore(driver_ip, run_func_server_port,
+                              'runfunc', 'func', pickled_exec_func)
+
+        command = [sys.executable, '-m', 'horovod.run.run_task', str(driver_ip), str(run_func_server_port)]
+
+        try:
+            _launch_job(args, remote_host_names, settings, common_intfs, command)
+            results = [None] * args.np
+            for i in range(args.np):
+                pickled_result = read_data_from_kvstore(driver_ip, run_func_server_port,
+                                                        'runfunc_result', str(i))
+                results[i] = cloudpickle.loads(pickled_result)
+            return results
+        finally:
+            run_func_server.shutdown_server()
+    else:
+        command = args.command
+        _launch_job(args, remote_host_names, settings, common_intfs, command)
+        return None
+
+
+def _launch_job(args, remote_host_names, settings, common_intfs, command):
     env = os.environ.copy()
     config_parser.set_env_from_args(env, args)
+    driver_ip = _get_driver_ip(common_intfs)
 
     if args.use_gloo:
         if not gloo_built(verbose=(settings.verbose >= 2)):
             raise ValueError('Gloo support has not been built.  If this is not expected, ensure CMake is installed '
                              'and reinstall Horovod with HOROVOD_WITH_GLOO=1 to debug the build error.')
-        gloo_run(settings, remote_host_names, common_intfs, env)
+        gloo_run(settings, remote_host_names, common_intfs, env, driver_ip, command)
     elif args.use_mpi:
         if not mpi_built(verbose=(settings.verbose >= 2)):
             raise ValueError('MPI support has not been built.  If this is not expected, ensure MPI is installed '
                              'and reinstall Horovod with HOROVOD_WITH_MPI=1 to debug the build error.')
-        mpi_run(settings, common_intfs, env)
+        mpi_run(settings, common_intfs, env, command)
     else:
         if mpi_built(verbose=(settings.verbose >= 2)):
-            mpi_run(settings, common_intfs, env)
+            mpi_run(settings, common_intfs, env, command)
         elif gloo_built(verbose=(settings.verbose >= 2)):
-            gloo_run(settings, remote_host_names, common_intfs, env)
+            gloo_run(settings, remote_host_names, common_intfs, env, driver_ip, command)
         else:
             raise ValueError('Neither MPI nor Gloo support has been built. Try reinstalling Horovod ensuring that '
                              'either MPI is installed (MPI) or CMake is installed (Gloo).')
 
 
+def run_commandline():
+    args = parse_args()
+    args.run_func = None
+    _run(args)
+
+
+def run(
+        func,
+        args=(),
+        kwargs={},
```

Comment:
Do not use `{}` as the default value because it is mutable. Use `None` instead.

~~~python
>>> def test(a={}):
...   a["b"] = 1
...   print(a)

>>> test()

>>> help(test)

test(a={'b': 1})
~~~